File: easy/array/python/189-rotate-array.py
from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        # store last k somewhere
        temp = [0] * n
        # move n-k elems to the right
        for i in range(0, n-k):
            temp[i+k] = nums[i]
        # move temp back
        for i in range(0, k):
            temp[i] = nums[i-k]

        # Copy back to nums
        for i in range(len(temp)):
            nums[i] = temp[i]


print(Solution().rotate([1, 2, 3, 4, 5, 6, 7], 1))
print(Solution().rotate([1, 2, 3, 4, 5, 6, 7], 2))
print(Solution().rotate([1, 2, 3, 4, 5, 6, 7], 3))

File: easy/array/cpp/189-rotate-array.cpp
#include <iostream>
#include <vector>

using std::vector;

template <class X>
void print_vector(vector<X> vect)
{
    for (X i : vect)
    {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}

class Solution
{
public:
    void rotate(vector<int> &nums, int k)
    {
        int n = nums.size();
        if (k > n)
        {
            k = k % n;
        }
        vector<int>
            temp(n);

        for (size_t i = 0; i < n - k; i++)
        {
            temp[i + k] = nums[i];
        }

        for (size_t i = 0; i < k; i++)
        {
            temp[i] = nums[n - k + i];
        }

        for (size_t i = 0; i < nums.size(); i++)
        {
            nums[i] = temp[i];
        }
    }
};

int main()
{
    Solution a;
    vector<int> vect{1, 2, 3, 4, 5, 6, 7};
    a.rotate(vect, 7);
    print_vector(vect);
}
